stan/math/rev/core/precomputed_gradients.hpp

```diff
@@ -96,10 +94,7 @@ class precomputed_gradients_vari : public vari {
  * @return An autodiff variable that uses the precomputed
  * gradients provided.
  */
-template <typename Arith, typename VecVar, typename VecArith,
-          require_arithmetic_t<Arith>...,
-          require_vector_like_vt<is_var, VecVar>...,
-          require_vector_like_vt<std::is_arithmetic, VecArith>...>
+template <typename Arith, typename VecVar, typename VecArith>
```

Comment:
Why are these gone? I think we only want to support arithmetic and vector like objects here? Or should this also take in complex types now?

Reply:
If it should take in complex as well then maybe we should have a `is_real` type trait to check if a type is arithmetic or complex with an arithmetic base type.

Reply:
They're gone because (a) they're causing circular include problems between core and meta, which I wasn't able to resolve, and (b) they're not needed.  The traits restrictions aren't necessary when there's no ambiguity to resolve in calling them.  If a client tries to call this code and the required concepts for the arguments (like supporting `operator[]`) aren't satisfied, the function won't compile at that juncture, rather than at the undefined-because-of-SFINAE level.

Reply:
And no, I wasn't thinking about `complex`.  

That's another whole layer of metaprogram reworking as you suggest.  I started in the last complex PR by splitting return traits into `real_return_t<T1, ..., TN>` and `complex_return_t<T1, ..., TN>`).  Having rhe `is_XXX` and `require_XXX` split is also going to be useful.

P.S.  Thanks for looking through this!

Reply:
Alright I'm fine with that and np!